### PR TITLE
Implement ranges::iota_view

### PR DESCRIPTION
**Pull request by CaseyCarter**
Also rework the implementation of P1739R4 to propose as a resolution for LWG-3407, and implement the resolution of LWG-3523.

### CODE REVIEW

File: stl/inc/ranges [hunk 1]
```diff
@@ -534,6 +535,332 @@ namespace ranges {
...
+        template <class _Left, class _Right>
+        static constexpr bool _Nothrow_minus_equal = noexcept(_STD declval<_Left&>() -= _STD declval<const _Right&>());
+        template <_Integer_like _Left, class _Right>
+        static constexpr bool _Nothrow_minus_equal<_Left, _Right> = true;
+
+    public:
+#endif // TRANSITION, DevCom-1347136
+
+        constexpr _Ioterator& operator+=(const difference_type _Off)
+#if defined(__clang__) || defined(__EDG__) // TRANSITION, DevCom-1347136
```
Comment (CaseyCarter) [inline stl/inc/ranges:624]: DevCom-1347136 is fun: the compiler warns about possible loss of data narrowing `_Off` in `noexcept(_Current += _Off)` (Yes, loss of data in an _unevaluated_ _operand_.)

File: stl/inc/ranges [hunk 2]
```diff
@@ -534,6 +535,332 @@ namespace ranges {
...
+            noexcept(noexcept(_Current += _Off)) /* strengthened */
+#else // ^^^ no workaround / workaround vvv
+            noexcept(noexcept(_Nothrow_plus_equal<_Wi, difference_type>)) /* strengthened */
```
Comment (cpplearner) [inline stl/inc/ranges:627]:
```suggestion
            noexcept(_Nothrow_plus_equal<_Wi, difference_type>) /* strengthened */

```

Comment (CaseyCarter) [reply stl/inc/ranges:627]: Ugh.

File: stl/inc/ranges [hunk 3]
```diff
@@ -534,6 +535,332 @@ namespace ranges {
...
+        /* [[no_unique_address]] */ _Bo _Last{};
+
+        _NODISCARD friend constexpr bool operator==(const _It& _Left, const _Iotinel& _Right) noexcept(
+            noexcept(_Right._Equal(_Left))) /* strengthened */ {
+            return _Right._Equal(_Left);
+        }
+
+        _NODISCARD friend constexpr iter_difference_t<_Wi> operator-(const _It& _Left, const _Iotinel& _Right) noexcept(
+            noexcept(_Right._Delta(_Left))) /* strengthened */ requires sized_sentinel_for<_Bo, _Wi> {
+            return -_Right._Delta(_Left);
```
Comment (StephanTLavavej) [inline stl/inc/ranges:751]: Do we have to worry about the unary negation potentially throwing?

File: stl/inc/ranges [hunk 4]
```diff
@@ -534,6 +535,332 @@ namespace ranges {
...
+
+    // clang-format off
+    template <class _Wi, class _Bo>
+        requires (!_Integer_like<_Wi> || !_Integer_like<_Bo>
+            || (_Signed_integer_like<_Wi> == _Signed_integer_like<_Bo>))
+    iota_view(_Wi, _Bo) -> iota_view<_Wi, _Bo>;
+    // clang-format on
+
+    template <class _Wi, class _Bo>
+    inline constexpr bool enable_borrowed_range<iota_view<_Wi, _Bo>> = true;
```
Comment (StephanTLavavej) [inline stl/inc/ranges:841]: I observe that WG21-N4878 [ranges.syn] depicts `template<weakly_incrementable W, semiregular Bound>` here. I don't think that the constraints matter, but wanted to double-check.

File: stl/inc/ranges [hunk 5]
```diff
@@ -1730,13 +2064,26 @@ namespace ranges {
...
                     // it's a "reconstructible range"; return the same kind of range with a restricted extent
                     _Count            = (_STD min)(_RANGES distance(_Range), _Count);
                     const auto _First = _RANGES begin(_Range);
-                    return remove_cvref_t<_Rng>{_First, _First + _Count};
-                } else if constexpr (_Strat == _St::_Take_view) {
-                    return take_view{_STD forward<_Rng>(_Range), _Count};
+
+                    // The following are all per the proposed resolution of LWG-3407
+                    if constexpr (_Strat == _St::_Reconstruct_span) {
+                        return span(_First, _First + _Count);
```
Comment (StephanTLavavej) [inline stl/inc/ranges:2076]: Can we use braces to construct this temporary?
```suggestion
                        return span{_First, _First + _Count};
```

File: stl/inc/ranges [hunk 6]
```diff
@@ -2053,18 +2400,32 @@ namespace ranges {
...
+                                _RANGES end(_STD declval<_Rng&>())})};
+                    } else {
+                        return {_St::_Reconstruct_subrange,
+                            noexcept(_Ty{_RANGES begin(_STD declval<_Rng&>()) + _RANGES distance(_STD declval<_Rng&>()),
+                                _RANGES end(_STD declval<_Rng&>()), range_difference_t<_Rng>{0}})};
+                    }
+                } else if constexpr (_Random_sized_range<_Ty> && _Is_specialization_v<_Ty, iota_view>) {
+                    return {_St::_Reconstruct_other,
+                        noexcept(_Ty(_RANGES begin(_STD declval<_Rng&>()) + _RANGES distance(_STD declval<_Rng&>()),
+                            _RANGES end(_STD declval<_Rng&>())))};
```
Comment (StephanTLavavej) [inline stl/inc/ranges:2428]: Nitpick: This says `_Ty(` but can it say `_Ty{` like the lines above?

File: tests/std/tests/P0896R4_views_iota/test.cpp [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    constexpr T low        = 0;
+    constexpr T high       = 8;
+    constexpr T expected[] = {0, 1, 2, 3, 4, 5, 6, 7};
+
+    {
+        // Validate bounded (both upper and lower bounds exist) iota_view
+        using R = ranges::iota_view<T, T>;
+
+        // Validate type properties
+        static_assert(same_as<ranges::range_value_t<R>, T>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_iota/test.cpp:33]: Include `<concepts>` for `same_as`?

File: tests/std/tests/P0896R4_views_iota/test.cpp [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+        static_assert(same_as<ranges::range_reference_t<R>, T>);
+
+        static_assert(ranges::random_access_range<R>);
+        static_assert(!ranges::contiguous_range<R>);
+        static_assert(ranges::common_range<R>);
+
+        static_assert(ranges::view<R>);
+        static_assert(semiregular<R>);
+        static_assert(is_nothrow_copy_constructible_v<R>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_iota/test.cpp:42]: Need `<type_traits>`.

File: tests/std/tests/P0896R4_views_iota/test.cpp [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+    static_assert((test_integral<char32_t>(), true));
+    test_integral<char32_t>();
+
+    // Some non-integer coverage:
+    {
+        // Pointers
+        empty_type objects[] = {{}, {}, {}};
+        const auto address   = [](auto& x) { return &x; };
+        assert(ranges::equal(
+            views::iota(begin(as_const(objects)), end(objects)), objects, ranges::equal_to{}, identity{}, address));
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_iota/test.cpp:277]: Need `<utility>` for `as_const()`.

Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_iota/test.cpp:277]: Need `<functional>` for `identity`.

File: tests/std/tests/P0896R4_views_iota/test.cpp [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+
+        assert(first == I{T{0}});
+        assert(second == I{T{1}});
+        static_assert(noexcept(noexcept(I{T{0}}))); // strengthened
+
+        {
+            I tmp = first;
+            static_assert(is_nothrow_copy_constructible_v<I>);
+
+            I i = std::move(tmp);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_iota/test.cpp:101]: This test is already `using namespace std;`. Occurs below.

File: tests/std/tests/P0896R4_views_iota/test.cpp [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+            static_assert(same_as<ranges::range_difference_t<R>, int>);
+        } else {
+            static_assert(same_as<ranges::range_difference_t<R>, long long>);
+        }
+
+        {
+            const same_as<R> auto rng = views::iota(low);
+
+            auto i = low;
+            for (auto const& e : rng) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_iota/test.cpp:219]:
```suggestion
            for (const auto& e : rng) {
```

File: tests/std/tests/P0896R4_views_iota/test.cpp [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+    static_assert((test_integral<unsigned short>(), true));
+    test_integral<unsigned short>();
+    static_assert((test_integral<unsigned int>(), true));
+    test_integral<unsigned int>();
+    static_assert((test_integral<unsigned long>(), true));
+    test_integral<unsigned long>();
+    static_assert((test_integral<unsigned long long>(), true));
+    test_integral<unsigned long long>();
+
+    // Validate other integer types
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_iota/test.cpp:256]: Also `wchar_t`?

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
Looks good, all I found were nitpicks!

File: stl/inc/ranges (hunk 3 above)
Comment (miscco) [reply stl/inc/ranges:751]: The horror

File: tests/std/tests/P0896R4_views_iota/test.cpp (hunk 11 above)
Comment (SuperWig) [reply tests/std/tests/P0896R4_views_iota/test.cpp:219]: I thought you said nitpicks, Stephan.

File: stl/inc/ranges [hunk 13]
```diff
@@ -534,6 +535,332 @@ namespace ranges {
...
+            return _Current;
+        }
+
+        constexpr _Ioterator& operator++() noexcept(noexcept(++_Current)) /* strengthened */ {
+            ++_Current;
+            return *this;
+        }
+
+        constexpr auto operator++(int) noexcept(
+            is_nothrow_copy_constructible_v<_Wi>&& noexcept(++_Current)) /* strengthened */ {
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges:585]: I've been making an effort to keep conditional `noexcept` simple, even to the extent of ignoring terms like `is_nothrow_copy_constructible_v` here which really only applies to one path of several. I'll tighten this up to `noexcept(++_Current) && (!incrementable<_Wi> || is_nothrow_copy_constructible_v<_Wi>)`, though since it bothers people.

File: stl/inc/ranges [hunk 14]
```diff
@@ -534,6 +535,332 @@ namespace ranges {
...
+            if constexpr (_Integer_like<_Wi>) {
+                if constexpr (!_Signed_integer_like<_Wi>) {
+                    if (_Off >= difference_type{0}) {
+                        _Current += static_cast<_Wi>(_Off);
+                    } else {
+                        _Current -= static_cast<_Wi>(-_Off);
+                    }
+                } else {
+                    _Current = static_cast<_Wi>(_Current + _Off);
+                }
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges:639]: Yes! I made a huge mess of these trying to find the source of the compiler warning from DevCom-1347136, and apparently didn't get them all the way cleaned up again.

File: stl/inc/ranges [hunk 15]
```diff
@@ -534,6 +535,332 @@ namespace ranges {
...
+            noexcept(noexcept(static_cast<_Wi>(_Current + _Idx))) /* strengthened */ requires _Advanceable<_Wi> {
+            if constexpr (_Integer_like<_Wi>) {
+                return static_cast<_Wi>(_Current + static_cast<_Wi>(_Idx));
+            } else {
+                return static_cast<_Wi>(_Current + _Idx);
+            }
+        }
+
+        _NODISCARD friend constexpr bool operator==(
+            const _Ioterator& _Left, const _Ioterator& _Right) requires equality_comparable<_Wi> = default;
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges:679]: There's no pasta like copy pasta!

File: stl/inc/ranges [hunk 16]
```diff
@@ -534,6 +535,332 @@ namespace ranges {
...
+            return _Right._Current < _Left._Current;
+        }
+        _NODISCARD friend constexpr bool operator<=(const _Ioterator& _Left, const _Ioterator& _Right) noexcept(
+            noexcept(_Left._Current < _Right._Current)) /* strengthened */ requires totally_ordered<_Wi> {
+            return !(_Right._Current < _Left._Current);
+        }
+        _NODISCARD friend constexpr bool operator>=(const _Ioterator& _Left, const _Ioterator& _Right) noexcept(
+            noexcept(_Left._Current < _Right._Current)) /* strengthened */ requires totally_ordered<_Wi> {
+            return !(_Left._Current < _Right._Current);
+        }
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges:695]: I'm avoiding the marginal cost to determine that `!` on the `bool` that `meow < woof` yields isn't potentially-throwing, which probably isn't enough of a throughput win to justify even minimal confusion.

File: stl/inc/ranges [hunk 17]
```diff
@@ -534,6 +535,332 @@ namespace ranges {
...
+        _NODISCARD friend constexpr _Ioterator operator+(_Ioterator _It, const difference_type _Off) noexcept(
+            noexcept(static_cast<_Wi>(_It._Current + _Off))) /* strengthened*/ requires _Advanceable<_Wi> {
+            return _Ioterator{static_cast<_Wi>(_It._Current + _Off)};
+        }
+        _NODISCARD friend constexpr _Ioterator operator+(const difference_type _Off, _Ioterator _It) noexcept(
+            noexcept(static_cast<_Wi>(_It._Current + _Off))) /* strengthened*/ requires _Advanceable<_Wi> {
+            return _Ioterator{static_cast<_Wi>(_It._Current + _Off)};
+        }
+        _NODISCARD friend constexpr _Ioterator operator-(_Ioterator _It, const difference_type _Off) noexcept(
+            noexcept(static_cast<_Wi>(_It._Current + _Off))) /* strengthened*/ requires _Advanceable<_Wi> {
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges:713]: 🐞 squashed. Excellent catch.

File: stl/inc/ranges [hunk 18]
```diff
@@ -534,6 +535,332 @@ namespace ranges {
...
+        _NODISCARD friend constexpr difference_type
+            operator-(const _Ioterator& _Left, const _Ioterator& _Right) noexcept(
+                noexcept(_Left._Current - _Right._Current)) /* strengthened */ requires _Advanceable<_Wi> {
+            return static_cast<difference_type>(_Left._Current - _Right._Current);
+        }
+    };
+
+    // clang-format off
+    template <weakly_incrementable _Wi, semiregular _Bo>
+        requires _Weakly_equality_comparable_with<_Wi, _Bo> && semiregular<_Wi>
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges:725]: This is a comedy of errors. That term was added in one of the _four_ necessary places by LWG-3292. I noticed it was absent from one of the necessary places, which I fixed in cplusplus/draft#4192. @johelegp noticed it missing from the other two which he fixed in cplusplus/draft#4426, using the `\libconcept{semiregular}` markup that the editors have devised in an attempt to index every use of every concept. Sadly that change was missing the `@` markup to tell LaTeX to switch from code to text mode so it rendered literally:
image
which I fixed yesterday in cplusplus/draft#4503. Open source at work, folks!

TLDR: Yes, this is intentional.

File: stl/inc/ranges (hunk 3 above)
Comment (CaseyCarter) [reply stl/inc/ranges:751]: Recall that integer-like types - the types allowed for difference types of models of `weakly_incrementable` - are the union of the mutually exclusive sets of integer types and integer-class types. Integer-class types are an implementation-defined set of types (currently empty for us) that are specified to look and act like an integer type.

[[iterator.concept.winc]/5.1](http://eel.is/c++draft/iterator.concept.winc#5.1) says: 

> For every unary operator @ for which the expression @x is well-formed, @a shall also be well-formed and have the same value, effects, and value category as @x provided that value is representable by I.

The intent is that "effects" here means "doesn't throw since the operation on integers doesn't throw."

File: stl/inc/ranges (hunk 4 above)
Comment (CaseyCarter) [reply stl/inc/ranges:841]: if this specialization is triggered for `iota_view<T, U>`, we know that someone somewhere formed the _template-id_ `iota_view<T, U>. When they did, the compiler checked that `T` and `U` satisfy `iota_view`'s constraints ([[temp.names]/7](http://eel.is/c++draft/temp.names#7)). Consequently it's redundant to check those constraints again for cases like this where we're matching a specialization, the only effect is to increase the number of places we have to keep the set of constraints synchronized.

I'll see if I can convince the editors that removing constraints here and on `enable_borrowed_range<subrange<I, S, K>>` is editorial (cplusplus/draft#4519).

File: tests/std/tests/P0896R4_views_iota/test.cpp (hunk 7 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_views_iota/test.cpp:33]: The standard requires `<ranges>` to include `<iterator>` which includes `<concepts>`. Weirdly, it also requires both `<ranges>` and `<iterator>` to include `<compare>`. 🤷

File: tests/std/tests/P0896R4_views_iota/test.cpp (hunk 11 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_views_iota/test.cpp:219]: `const`-west is `const`-worst; east-`const` is beast-`const`!

File: stl/inc/ranges [hunk 19]
```diff
@@ -534,6 +535,332 @@ namespace ranges {
...
+            _Ty(__j + __n);
+            _Ty(__n + __j);
+            _Ty(__j - __n);
+            { __j - __j } -> convertible_to<_Iota_diff_t<_Ty>>;
+        };
+
+    // clang-format off
+    template <weakly_incrementable _Wi>
+        requires semiregular<_Wi>
+    struct _Ioterator {
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges:564]: Since the iterator doesn't depend on the sentinel type, this allows `_Ioterator<T>` to be reused for both `iota_view<T, U>` and `iota_view<T, V>`. (See ["What Are SCARY Iterators?"](https://devblogs.microsoft.com/cppblog/what-are-scary-iterators/) for details on this practice.)

[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/ranges:563]: That is aweome

File: tests/std/tests/P0896R4_views_iota/test.cpp [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+        // Bounded and unbounded iota_view have the same iterator type, however:
+        static_assert(same_as<ranges::iterator_t<ranges::iota_view<T, T>>, ranges::iterator_t<ranges::iota_view<T>>>);
+
+        if constexpr (sizeof(T) < sizeof(int)) {
+            static_assert(same_as<ranges::range_difference_t<R>, int>);
+        } else {
+            static_assert(same_as<ranges::range_difference_t<R>, long long>);
+        }
+
+        {
```
Comment (miscco) [inline tests/std/tests/P0896R4_views_iota/test.cpp:218]: Is there some condition missing here?

I do not really see the benefit of that scope, as there are no other variables defined here

File: stl/inc/ranges (hunk 19 above)
[the remark this answers is not in the document]
Comment (SuperWig) [reply stl/inc/ranges:563]: That acronym makes absolutely no sense. The letters are just randomly picked from the words!

[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/ranges:563]: I mean its C++. Our acronyms are even worse than our defaults

File: tests/std/tests/P0896R4_views_iota/test.cpp (hunk 20 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_views_iota/test.cpp:218]: Mostly the braces are here (as on other blocks) to wrap this up in a nicely self-contained chunk. It's probably my lizard brain telling me to factor this overlarge function out into separate small and simple functions. That said I acknowledge that removing this pair of braces reduces indentation but has no semantic effect. I can strike them if you think it's an improvement.

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/ranges [hunk 21]
```diff
@@ -534,6 +535,331 @@ namespace ranges {
...
+        _NODISCARD friend constexpr bool operator<=(const _Ioterator& _Left, const _Ioterator& _Right) noexcept(
+            noexcept(!(_Right._Current < _Left._Current))) /* strengthened */ requires totally_ordered<_Wi> {
+            return !(_Right._Current < _Left._Current);
+        }
+        _NODISCARD friend constexpr bool operator>=(const _Ioterator& _Left, const _Ioterator& _Right) noexcept(
+            noexcept(!(_Left._Current < _Right._Current))) /* strengthened */ requires totally_ordered<_Wi> {
+            return !(_Left._Current < _Right._Current);
+        }
+        // clang-format off
+        _NODISCARD friend constexpr auto operator<=>(const _Ioterator& _Left, const _Ioterator& _Right) noexcept(
```
Comment (MahmoudGSaleh) [inline stl/inc/ranges:696]: With <=>, why do we still need the other operators?

**Review by MahmoudGSaleh — APPROVED** [review]

File: stl/inc/ranges (hunk 21 above)
Comment (CaseyCarter) [reply stl/inc/ranges:696]: `_Ioterator`'s operations are implemented in terms of the operations of `_Wi`, so `_Ioterator`'s `<=>` requires `_Wi` to support `<=>`, but `_Ioterator`'s `<` only requires `_Wi` to support `<`. This way the other operations still work on `_Ioterator`s even when the wrapped type doesn't implement `<=>`.

**Comment by StephanTLavavej** [comment]
1, 2, 3, 4, thank you for adding more, 5, 6, 7, 8, views in `<ranges>` are super great! :heart_eyes_cat: :rocket: :1234:
